utils/s3/client.cc

```diff
@@ -173,9 +173,13 @@ future<> client::make_request(http::request req, http::experimental::client::rep
     auto it = _https.find(sg);
     if (it == _https.end()) [[unlikely]] {
         auto factory = std::make_unique<dns_connection_factory>(_host, _cfg->port, _cfg->use_https);
+        // Limit the maximum number of connections this group's http client
+        // may have proportional to its shares. Shares are typically in the
+        // range of 100...1000, thus resulting in 1..10 connections
+        auto max_connections = std::max((unsigned)(sg.get_shares() / 100), 1u);
         it = _https.emplace(std::piecewise_construct,
             std::forward_as_tuple(sg),
-            std::forward_as_tuple(std::move(factory))
+            std::forward_as_tuple(std::move(factory), max_connections)
         ).first;
     }
```

Comment:
by the way, shares can be dynamic (specially one for memtable flush and compaction), so it's important to update the maximum # of connection for respective sched groups. the good news is that http client::set_maximum_connections can shrink or grow the number of connections. I assume shrink waits for ongoing activity in the outstanding connections being removed.

I'd suggest introducing a fast path in set_maximum_connections(), to bail out if nr == _max_connections. Wait, it does have.

Reply:
> by the way, shares can be dynamic (specially one for memtable flush and compaction), so it's important to update the maximum # of connection for respective sched groups.

Of course, but it's heavier patching, so it's not yet here

> I'd suggest introducing a fast path in set_maximum_connections(), to bail out if nr == _max_connections. Wait, it does have.

Yes, it 'does nothing" if nothing changes :)